### PR TITLE
ENH - Add ``Pinball`` datafit

### LINKED ISSUE #133: FEAT add Pinball loss
Aptly suggested by @Badr-MOUFAD

### PR DESCRIPTION
**Pull request by Badr-MOUFAD**
Provided that we now have a solver that handles non-smooth datafits  #131

This adds a Pinball datafit that can be used to fit a ``QuantileRegressor`` 
and hence as a by-product fit a LAD-Lasso (``quantile=0.5``)

closes #133

### CODE REVIEW

File: skglm/experimental/pdcd_ws.py [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+import numpy as np
+from numpy.linalg import norm
+from scipy.sparse import issparse
+
+from numba import njit
+from skglm.utils.jit_compilation import compiled_clone
+from sklearn.exceptions import ConvergenceWarning
+
+
+class PDCD_WS:
```
Comment (mathurinm) [inline skglm/experimental/pdcd_ws.py:12]: you need to merge upstream/main into your PR branch for these to disappear from the diff

File: skglm/experimental/quantile_regression.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class Pinball(BaseDatafit):
+    r"""Pinball datafit.
+
+    The datafit reads::
+
+        quantile * max(y - Xw, 0) + (1 - quantile) * max(Xw - y, 0)
+
+    such that quantile in ``[0, 1]``.
```
Comment (mathurinm) [inline skglm/experimental/quantile_regression.py:14]:
```suggestion
    with ``quantile`` in [0, 1].
```

File: skglm/experimental/quantile_regression.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        quantile = self.quantile
+
+        residual = y - Xw
+        sign = residual >= 0
+
+        loss = quantile * sign * residual - (1 - quantile) * (1 - sign) * residual
+        return np.sum(loss)
+
+    def prox(self, w, step, y):
+        """Prox of ||y - . || with step ``step``."""
```
Comment (mathurinm) [inline skglm/experimental/quantile_regression.py:38]: to update

File: skglm/experimental/quantile_regression.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        shift_cst = (self.quantile - 1/2) * step
+        return y - ST_vec(y - w - shift_cst, step / 2)
+
+    def prox_conjugate(self, z, step, y):
+        """Prox of ||y - . ||^* with step ``step``."""
```
Comment (mathurinm) [inline skglm/experimental/quantile_regression.py:43]: same

File: skglm/experimental/pdcd_ws.py (hunk 1 above)
Comment (Badr-MOUFAD) [reply skglm/experimental/pdcd_ws.py]: Yes indeed!

**Comment by Badr-MOUFAD** [comment]
The [benchmarks results](https://badr-moufad.github.io/share-benchmarks/quantile_regression/benchmark_quantile_regression_benchopt_run_2022-12-08_13h36m59.html) for the following setups:
- datasets: MEG, Simulated
- lambda: ``[0.5, 0.1, 0.05] * lambda_max``
- solvers: ``PDCD-WS``, ``scipy.linprog`` solvers

@lorentzenchr, your feedback is valuable :pray:

File: skglm/experimental/quantile_regression.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class Pinball(BaseDatafit):
+    r"""Pinball datafit.
+
+    The datafit reads::
+
+        quantile * max(y - Xw, 0) + (1 - quantile) * max(Xw - y, 0)
+
+    such that ``quantile`` in [0, 1].
```
Comment (mathurinm) [inline skglm/experimental/quantile_regression.py:14]:
```suggestion
    with ``quantile`` in [0, 1].
```

File: skglm/experimental/quantile_regression.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+from skglm.datafits import BaseDatafit
+from skglm.utils.prox_funcs import ST_vec
+
+
+class Pinball(BaseDatafit):
+    r"""Pinball datafit.
+
+    The datafit reads::
+
+        quantile * max(y - Xw, 0) + (1 - quantile) * max(Xw - y, 0)
```
Comment (mathurinm) [inline skglm/experimental/quantile_regression.py:12]: nitpick: the real value does not involve np.max, it is np.maximum(...). sum().
Maybe rewrite as a sum and use _i to denote sample indices ? check how sklearn does it

Comment (Badr-MOUFAD) [reply skglm/experimental/quantile_regression.py:12]: I looked up [scikit-learn source code](https://github.com/scikit-learn/scikit-learn/blob/dc580a8ef5ee2a8aea80498388690e2213118efd/sklearn/linear_model/_quantile.py#L181), but they don't specify the expression.

I am more with the usage of ``sum`` and ``_`` to indicate samples.

File: skglm/experimental/quantile_regression.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+from skglm.datafits import BaseDatafit
+from skglm.utils.prox_funcs import ST_vec
+
+
+class Pinball(BaseDatafit):
+    r"""Pinball datafit.
+
+    The datafit reads::
+
+        sum_i quantile * max(y_i - Xw_i, 0) + (1 - quantile) * max(Xw_i - y_i, 0)
```
Comment (lorentzenchr) [inline skglm/experimental/quantile_regression.py:12]: The naming "quantile" is unfortunate. It should be "quantile level", i.e. the quantile at level 50% is the median. The "quantile" in the formula is not the quantile, but Xw_i is an estimation of the quantile.

**Review by lorentzenchr — COMMENTED** [review]
I can't comment on the core functions like `prox` as I'm not too familiar with skglm.
A few more tests could not hurt.

File: skglm/experimental/quantile_regression.py (hunk 7 above)
Comment (mathurinm) [reply skglm/experimental/quantile_regression.py:12]: thanks @lorentzenchr for the insightful remark, we'll change that

**Review by mathurinm — APPROVED** [review]
